File: templates/problems/07_trees/11_validate_bst/validate_bst_test.go
package validate_bst

import "testing"

func TestIsValidBST(t *testing.T) {
	tests := []struct {
		name string
		vals []int
		want bool
	}{
		{"valid bst", []int{2, 1, 3}, true},
		{"invalid bst", []int{5, 1, 4, -101, -101, 3, 6}, false},
		{"single node", []int{1}, true},
		{"empty tree", []int{}, true},
		{"left equal", []int{2, 2, 3}, false},
		{"valid larger", []int{5, 3, 7, 1, 4, 6, 8}, true},
		{"invalid deep right", []int{5, 4, 6, -101, -101, 3, 7}, false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			root := buildTreeHelper(tt.vals)
			if got := isValidBST(root); got != tt.want {
				t.Errorf("isValidBST() = %v, want %v", got, tt.want)
			}
		})
	}
}

File: templates/problems/07_trees/11_validate_bst/typescript/validate_bst.test.ts
import { describe, it, expect } from "vitest";
import { buildTree } from "./helpers";
import { isValidBST } from "./validate_bst";

describe("isValidBST", () => {
    it.each([
        { name: "valid bst", vals: [2, 1, 3], want: true },
        { name: "invalid bst", vals: [5, 1, 4, -101, -101, 3, 6], want: false },
        { name: "single node", vals: [1], want: true },
        { name: "empty tree", vals: [], want: true },
        { name: "left equal", vals: [2, 2, 3], want: false },
        { name: "valid larger", vals: [5, 3, 7, 1, 4, 6, 8], want: true },
        { name: "invalid deep right", vals: [5, 4, 6, -101, -101, 3, 7], want: false },
    ])("case: $name", ({ vals, want }) => {
        const root = buildTree(vals);
        expect(isValidBST(root)).toBe(want);
    });
});
